### PR TITLE
Initial support for ProcessGroups

### PR DESCRIPTION
**Pull request by johscheuer**
Fixes: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/410

This PR adds the `ProcessGroups` proposed in https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/design/process_group_status.md. Currently the status is not used by the reconciler and will only be filled in in the `update_status` reconciler. The next step is to actually use the fields instead of the deprecated fields.

### CODE REVIEW

**Comment by johscheuer** [comment]
The resulting status for my test cluster:

```yaml   
 processGroups:
    - addresses:
      - 10.1.1.193
      grocessGroupID: cluster_controller-1
      processClass: cluster_controller
    - addresses:
      - 10.1.1.195
      grocessGroupID: log-1
      processClass: log
    - addresses:
      - 10.1.1.198
      grocessGroupID: log-2
      processClass: log
    - addresses:
      - 10.1.1.196
      grocessGroupID: log-3
      processClass: log
    - addresses:
      - 10.1.1.194
      grocessGroupID: log-4
      processClass: log
    - addresses:
      - 10.1.1.192
      grocessGroupID: storage-1
      processClass: storage
    - addresses:
      - 10.1.1.191
      grocessGroupID: storage-2
      processClass: storage
    - addresses:
      - 10.1.1.197
      grocessGroupID: storage-3
      processClass: storage
```

File: api/v1beta1/foundationdbcluster_types.go [hunk 1]
```diff
@@ -340,8 +346,190 @@ type FoundationDBClusterStatus struct {
...
+
+	// ProcessGroups contain information about a process group.
+	// This information is used in multiple places to trigger the according action.
+	ProcessGroups []*ProcessGroupStatus `json:"processGroups,omitempty"`
+}
+
+// ProcessGroupStatus represents a the status of a ProcessGroup.
+type ProcessGroupStatus struct {
+	// ProcessGroupID represents the ID of the process group
+	ProcessGroupID string `json:"grocessGroupID,omitempty"`
```
Comment (brownleej) [inline api/v1beta1/foundationdbcluster_types.go:358]:
```suggestion
	ProcessGroupID string `json:"processGroupID,omitempty"`
```

**Review by brownleej — CHANGES_REQUESTED** [review]
Found a typo. Otherwise LGTM

File: api/v1beta1/foundationdbcluster_types.go (hunk 1 above)
Comment (johscheuer) [reply api/v1beta1/foundationdbcluster_types.go]: 🤦 Thanks!

**Comment by johscheuer** [comment]
Rebased for the kustomize fix.

**Review by brownleej — APPROVED** [review]
👍

File: controllers/update_status.go [hunk 2]
```diff
@@ -360,14 +362,18 @@ func (s UpdateStatus) Reconcile(r *FoundationDBClusterReconciler, context ctx.Co
...
 	_, err = cluster.CheckReconciliation()
 	if err != nil {
 		return false, err
 	}
 
-	if !reflect.DeepEqual(cluster.Status, *originalStatus) {
+	// See: https://github.com/kubernetes-sigs/kubebuilder/issues/592
+	// If we use the default reflect.DeepEqual method it will be recreating the
+	// status multiple times because the pointers are different.
+	if !equality.Semantic.DeepEqual(cluster.Status, *originalStatus) {
```
Comment (johscheuer) [inline controllers/update_status.go:376]: @brownleej I changed this to be robust against different pointers but the same content `reflect.DeepEqual` will check if the pointers are the same and `equality.Semantic.DeepEqual` will check if the content is the same which makes more sense in our case.

**Comment by brownleej** [comment]
This looks good. Can you take a look at the ProcessGroupID JSON tag again, and clean up the merge conflicts?

**Review by brownleej — APPROVED** [review]
This looks good. Can you take a look at the ProcessGroupID JSON tag again, and clean up the merge conflicts?

**Review by brownleej — APPROVED** [review]
👍

**Pull request merged**